File: leetcode/49. Group Anagrams/main.java
// Solved by @kitanoyoru
// https://leetcode.com/problems/group-anagrams/

import java.util.*;

class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
    if (strs.length == 0) {
      return new ArrayList<>();
    }

    HashMap<String, ArrayList<String>> hm = new HashMap<>();

    for (String s : strs) {
      char[] sarr = s.toCharArray();
      Arrays.sort(sarr);
      String temp = String.valueOf(sarr);

      if (hm.containsKey(temp)) {
        hm.get(temp).add(s);
      } else {
        ArrayList<String> l = new ArrayList<>();
        l.add(s);
        hm.put(temp, l);
      }
    }

    return new ArrayList<>(hm.values());
  }
}

File: leetcode/49. Group Anagrams/main.go
package main

import "sort"

func sortString(str string) string {
	runes := []rune(str)

	sort.Slice(runes, func(i, j int) bool {
		return runes[i] < runes[j]
	})

	return string(runes)
}

func groupAnagrams(strs []string) [][]string {
	if len(strs) == 0 {
		return [][]string{}
	}

	m := make(map[string][]string, len(strs))

	for _, s := range strs {
		key := sortString(s)

		if _, has := m[key]; has {
			m[key] = append(m[key], s)
		} else {
			m[key] = []string{s}
		}
	}

	values := make([][]string, 0, len(m))
	for _, val := range m {
		values = append(values, val)
	}

	return values
}
